osu.Game/Graphics/Containers/SectionsContainer.cs

```diff
@@ -195,16 +201,35 @@ protected override void UpdateAfterChildren()
                 headerBackgroundContainer.Y = ExpandableHeader?.Y ?? 0;
 
                 float scrollOffset = FixedHeader?.LayoutSize.Y ?? 0;
-                Func<T, float> diff = section => scrollContainer.GetChildPosInContent(section) - currentScroll - scrollOffset;
 
-                if (scrollContainer.IsScrolledToEnd())
+                float visibleSize(T section)
+                {
+                    float sectionPosition = scrollContainer.GetChildPosInContent(section) - scrollOffset;
+                    float top = Math.Max(sectionPosition, currentScroll);
+                    float bottom = Math.Min(sectionPosition + section.Height, currentScroll + scrollContainer.DisplayableContent);
+                    return bottom - top;
+                }
+
+                var sortedByVisible = new List<T>(Children);
+                sortedByVisible.Sort((a, b) => visibleSize(b).CompareTo(visibleSize(a)));
+                T mostVisible = sortedByVisible.FirstOrDefault();
+
+                if (mostVisible == null)
+                    return;
+
+                int mostVisibleIndex = IndexOf(mostVisible);
+                T ch;
+
+                while (mostVisibleIndex != 0 && Precision.AlmostEquals(visibleSize(ch = Children[mostVisibleIndex - 1]), ch.Height, 1))
```

Comment:
Probably want something like this so scrolling upwards using buttons feels as fast as scrolling downwards

```suggestion
                while (mostVisibleIndex != 0 && Precision.AlmostBigger(visibleSize(
                        ch = Children[mostVisibleIndex - 1]), ch.Height * (externallyScrolledTo as T == ch ? 0.8 : 1), 1
                ))
```

Reply:
I can't understand what this code snippet is proposing. What is `0.8` and why is it there?

Reply:
When scrolling downwards the section comes visible sooner. That's because our target position for the section is the top of the screen. When scrolling upwards, the section stays behind the header and isn't considered selected until it is 100% visible. Just thought that maybe we should make an exception for the section user clicked for and consider it selected when only 80% is visible. 80% was chosen arbitrarily, felt like good number.

Another fix for it would be to scroll the section to the middle of the screen. More specifically, instead of scrolling top of the section to top of the screen, scroll the middle of the section (or at minimum half the screen height if section takes up more than screen worth of space, just so section header stays on screen) to the middle of the screen.

Reply:
What about the exemption where it's 1 instead of 0.8? Also this assignment to `ch` within the loop is confusing me intensely. I'd like this loop to be simple, stupid code even, if possible.

Reply:
> What about the exemption where it's 1 instead of 0.8?

It's actually not an exemption, it's the default for every section except the one user wanted to see.

I would suggest just scrolling the section further than just he top of the screen, but then it becomes the question about how much further. Main issue would be for small sections that would become unselected if scroll the the middle, and like a mentioned sections that take up more space than screen height.

> I'd like this loop to be simple, stupid code even, if possible.

Isn't accessing an index O(1)? I agree that I should have just written ``Children[mostVisibleIndex - 1]`` more than once.

Reply:
> It's actually not an exemption, it's the default for every section except the one user wanted to see.

I'd say that should be accurately commented, then. It's not immediately obvious what that condition does. Maybe the other changes around this logic will improve this, not sure.

> Isn't accessing an index O(1)? I agree that I should have just written `Children[mostVisibleIndex - 1]` more than once.

Computational complexity is not what I'm concerned about here - code complexity is. This one-liner while condition has a test in it, followed by another test, that has a nested assignment and a function call. It's not easy reading. That can all be in the body and be several lines.